File: src/metri/janus/multi_series.clj
;; [PORTED_TO_RUST: src/janus/multi_series.rs]
;; NO MODIFICAR ESTE ARCHIVO.
;; La fuente de verdad para esta lógica ahora reside en Rust.
(ns metri.janus.multi-series
  "Motor de fusión MultiSeriesGroup — Full Outer Join asintótico.
   SRP: realiza y fusiona resultados de sub-queries agrupados bajo un group-id.

   Principios:
     • DIP: recibe process-query-fn como parámetro — no depende de core.clj.
     • Sin conocimiento de Cedar, AST compiler o Aegis.
     • La unión de filas es un Full Outer Join por hash de la row completa."
  (:require [taoensso.timbre :as log]
            [metri.otel.spans :as otel]))

;; ═══════════════════════════════════════════════════════════════════════════
;; FULL OUTER JOIN — fusión de filas multi-query
;; ═══════════════════════════════════════════════════════════════════════════

(defn- outer-join-data
  "Full Outer Join de filas de múltiples sub-queries.

   Para cada combinación única de valores de dimensión (representada como hash
   de la row completa), combina las métricas de todos los sub-queries en una
   única fila mediante merge de mapas.

   Filas sin contraparte en otros sub-queries reciben nil implícito para las
   métricas ausentes (Clojure map merge semántica)."
  [rows-per-qk]
  (if (= 1 (count rows-per-qk))
    (val (first rows-per-qk))
    (vals
      (reduce-kv
        (fn [idx _qk rows]
          (reduce (fn [i row]
                    (update i (hash row) #(merge (or % {}) row)))
                  idx rows))
        {}
        rows-per-qk))))

(defn- realize-query-data
  "Realiza los chunks de un sub-query y extrae las filas de datos.
   Descarta chunks de error — el Outer Join omite silenciosamente sub-queries fallidos."
  [chunks]
  (into []
        (comp (filter #(= :ok (first %)))
              (mapcat #(or (:data (second %)) [])))
        chunks))

;; ═══════════════════════════════════════════════════════════════════════════
;; PARTICIÓN — standalone vs agrupadas
;; ═══════════════════════════════════════════════════════════════════════════

(defn partition-standalone
  "Retorna el subconjunto de queries que NO pertenecen a ningún merge-group.
   Si merge-groups está vacío, retorna queries completo (O(1))."
  [queries merge-groups]
  (if (empty? merge-groups)
    queries
    (let [merged-qks (into #{} (mapcat :query-keys merge-groups))]
      (reduce-kv
        (fn [m qk qm]
          (if (contains? merged-qks (name qk)) m (assoc m qk qm)))
        {}
        queries))))

;; ═══════════════════════════════════════════════════════════════════════════
;; API PÚBLICA
;; ═══════════════════════════════════════════════════════════════════════════

(defn process-merge-groups
  "Ejecuta y fusiona los MultiSeriesGroup mediante Full Outer Join.

   Para cada grupo:
     1. Ejecuta cada sub-query vía process-query-fn (DIP — fn inyectada).
     2. Realiza los chunks → extrae filas de datos.
     3. Full Outer Join de filas de todos los sub-queries del grupo.
     4. Emite un chunk [:ok {:data joined :query-key (keyword group-id) ...}].

   process-query-fn: (fn [query-key query-map]) → seq de chunks
   tenant-id:        decorado en el chunk resultante para el cliente."
  [merge-groups queries process-query-fn tenant-id]
  (mapcat
    (fn [{:keys [group-id query-keys override-viz]}]
      (otel/with-span [(str "janus.merge-group." group-id) {:kind :internal}]
        (let [rows-per-qk
              (reduce
                (fn [acc qk-str]
                  (let [qk (keyword qk-str)
                        qm (get queries qk)]
                    (if (nil? qm)
                      (do (log/warn "[Janus] MultiSeriesGroup: query-key no encontrado"
                                    qk-str "| group:" group-id)
                          acc)
                      (assoc acc qk
                             (realize-query-data (process-query-fn qk qm))))))
                {}
                (or query-keys []))
              joined-data (outer-join-data rows-per-qk)]
          [[:ok {:data         joined-data
                 :query-key    (keyword group-id)
                 :group-id     group-id
                 :override-viz override-viz
                 :merged-keys  (mapv keyword (or query-keys []))
                 :tenant-id    tenant-id}]])))
    merge-groups))

File: src/janus/multi_series.rs
// [PORTED_FROM: src/metri/janus/multi_series.clj]
// janus/multi_series.rs — Motor de fusión MultiSeriesGroup — Full Outer Join asintótico.
// SRP: realiza y fusiona resultados de sub-queries agrupados bajo un group-id.

use serde_json::{json, Value, Map};
use std::collections::{HashMap, HashSet};
use std::hash::{Hash, Hasher};
use std::collections::hash_map::DefaultHasher;
use tracing::warn;

use crate::janus::router::QueryChunk;

/// Calcula un hash simplificado para un objeto JSON.
fn hash_value(val: &Value) -> u64 {
    let mut hasher = DefaultHasher::new();
    match val {
        Value::Object(map) => {
            // Sort keys to ensure consistent hashing
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            for k in keys {
                k.hash(&mut hasher);
                hash_value(map.get(k).unwrap()).hash(&mut hasher);
            }
        }
        Value::Array(arr) => {
            for v in arr {
                hash_value(v).hash(&mut hasher);
            }
        }
        Value::String(s) => s.hash(&mut hasher),
        Value::Number(n) => n.to_string().hash(&mut hasher),
        Value::Bool(b) => b.hash(&mut hasher),
        Value::Null => 0.hash(&mut hasher),
    }
    hasher.finish()
}

/// Full Outer Join de filas de múltiples sub-queries.
/// [PORTED_FROM: (outer-join-data rows-per-qk)]
pub fn outer_join_data(rows_per_qk: HashMap<String, Vec<Value>>) -> Vec<Value> {
    if rows_per_qk.len() == 1 {
        return rows_per_qk.into_values().next().unwrap();
    }

    let mut idx: HashMap<u64, Map<String, Value>> = HashMap::new();

    for (_, rows) in rows_per_qk {
        for row in rows {
            if let Value::Object(row_map) = row {
                let h = hash_value(&Value::Object(row_map.clone()));
                let entry = idx.entry(h).or_insert_with(Map::new);
                // Merge
                for (k, v) in row_map {
                    entry.insert(k, v);
                }
            }
        }
    }

    idx.into_values().map(Value::Object).collect()
}

/// Retorna el subconjunto de queries que NO pertenecen a ningún merge-group.
/// [PORTED_FROM: (partition-standalone queries merge-groups)]
pub fn partition_standalone(mut queries: Map<String, Value>, merge_groups: &[Value]) -> Map<String, Value> {
    if merge_groups.is_empty() {
        return queries;
    }

    let mut merged_qks = HashSet::new();
    for mg in merge_groups {
        if let Some(qks) = mg.get("query_keys").or(mg.get("query-keys")).and_then(|v| v.as_array()) {
            for qk in qks {
                if let Some(s) = qk.as_str() {
                    merged_qks.insert(s.to_string());
                }
            }
        }
    }

    let mut standalone = Map::new();
    for (k, v) in queries {
        if !merged_qks.contains(&k) {
            standalone.insert(k, v);
        }
    }

    standalone
}

// Nota: process-merge-groups se implementa en janus/router.rs directamente
// debido a las restricciones de async y el préstamo de closures en Rust.
